Docs: Add data-structure.md (#8039)
treeverse/lakeFS: docs/understand/data-structure.md

---
title: Data Structure
parent: Understanding lakeFS
description: Understand the data structure in lakeFS
---

# How Does lakeFS Store Your Data
lakeFS being a data versioning engine, requires the ability to save multiple versions of the same object. As a result, lakeFS stores objects in the object store
in an unconventional way.
This might cause confusion when trying to understand where our data is actually being stored. This page will try to shed a light on this subject.

## lakeFS Repository Namespace Structure

> lakeFS stores repository data and metadata under the repository's namespace.

Reviewer: Explain that a storage namespace is a path in an object store that's dedicated to the operation of that repository

Listing a repository storage namespace in the object store will provide the following output:

```shell
aws s3 ls s3://<storage_namespace>/
                        PRE _lakefs/
                        PRE data/
```

lakeFS stores the actual user data under the `data/` prefix. The `_lakefs/` prefix is used to store commit metadata which includes [range and meta-range](../understand/how/versioning-internals.md) files and internal lakeFS data.

> Since lakeFS manages immutable data, objects are not saved using their logical name. This means that when you upload a csv file called allstar_games_stats.csv to branch main, lakeFS will generate a random physical

Reviewer: ```suggestion
Since lakeFS manages immutable data, objects are not saved using their logical name - these might get overwritten, violating the immutability guarantee. This means that when you upload a csv file called `allstar_games_stats.csv` to branch main, lakeFS will generate a random physical
```

address under the `data/` prefix and upload it to there.  

> Mapping from a path to an object changes as you upload, commit, and merge on lakeFS. When updating an object, lakeFS will create a new physical address for that version preserving the different version of that object.

Reviewer: ```suggestion
Mapping from a path to an object changes as you upload, commit, and merge on lakeFS. When updating an object, lakeFS will create a new physical address for that version preserving other versions of that object.
```

> lakeFS will link between the objects logical address and physical address and store that relation under the given commit metadata (range and meta-range)

Reviewer: ```suggestion
lakeFS will link between the object's logical address and its physical address - and store that relation under the given commit metadata (range and meta-range)
```

Author: Already did that in line 22. Do you want me to add the link also here?

lakeFS uses its object store immutably i.e. anything uploaded is never changed or overridden (Refer to [GC](../howto/garbage-collection/index.md) for explanation on how and when lakeFS actually deletes data from the storage).  
To find data, lakeFS uses the logical address e.g. `lakefs://my-repo/main/allstar_games_stats.csv`, indicating a repository and branch. Graveler, the objects and versioning component, searches in several places in the following order:

> 1. In the KV metadata store under the current staging token of branch main. This will return any uncommitted changes for the given object

Reviewer: Not sure what this sentence means

Author: Refactored

> 2. Read it from the branch's head meta-range and range (which are saved under the `_lakefs` prefix in the object store. This will return the metadata for the object as it was stored in the latest commit for branch main.  

Author: Done
